File: services/commons/src/main/java/app/appgrove/commons/access/AppRole.java
package app.appgrove.commons.access;

import java.util.Optional;

/**
 * Ruolo di una persona <b>su una applicazione</b> (UC 0098). È il livello in cui il potere ha senso in
 * un modello centralizzato: il ruolo di piattaforma dice soltanto chi possiede l'account
 * ({@code MembershipRole} in core), mentre <i>cosa</i> una persona può fare lo dice questo.
 *
 * <p><b>Perché vive in {@code commons} e non in core</b> (UC 0099): il varco dichiarativo dei servizi
 * ({@link RequiresAppRole}) deve poter <b>nominare</b> il ruolo minimo richiesto, e {@code commons} è
 * l'unico posto che sia core sia i servizi delle applicazioni possono vedere ({@code commons} non
 * dipende da core, e non può). Duplicare l'enumerazione avrebbe duplicato l'<b>ordinamento</b>, che è
 * la cosa che meno di ogni altra va scritta due volte.
 *
 * <p>I tre valori sono <b>ordinati</b>, dal meno al più capace, e l'ordinamento è la dichiarazione
 * dell'ordine: {@code viewer} &lt; {@code editor} &lt; {@code admin}. Il confronto vive qui in
 * {@link #atLeast(AppRole)} e <b>non va duplicato altrove</b> — lo usano il varco riusabile dei servizi
 * e le schermate (UC 0101, UC 0111). Una seconda copia dell'ordinamento è una seconda occasione di
 * sbagliarlo.
 *
 * <p>Il <b>significato di comportamento</b> dei tre ruoli — cosa esattamente può fare un {@code editor}
 * in una applicazione qualunque — è il contratto di UC 0101: qui c'è soltanto l'ordine.
 *
 * <p><b>L'owner dell'account non compare fra questi valori</b>, e non è una dimenticanza: il suo accesso
 * è implicito su tutte le applicazioni dell'account e vale come {@link #admin} (UC 0098 §5). Sta sopra
 * tutti perché la fonte di verità gli attribuisce il valore massimo, non perché esista un quarto valore.
 */
public enum AppRole {

    /** Legge e non scrive. */
    viewer,

    /** Legge e modifica i dati dell'applicazione. */
    editor,

    /** Come {@code editor}, e in più governa gli accessi a <b>quella</b> applicazione (UC 0098 §2). */
    admin;

    /**
     * Questo ruolo è almeno tanto capace di quello richiesto? È il confronto che ogni operazione fa
     * prima di procedere: {@code caller.atLeast(EDITOR)}.
     */
    public boolean atLeast(AppRole required) {
        return ordinal() >= required.ordinal();
    }

    /**
     * Il ruolo scritto nel dato, o vuoto se il valore è assente o non riconosciuto. Un valore ignoto
     * <b>non</b> è un errore da propagare: è un'informazione che non sappiamo interpretare, e in
     * assenza di informazione non si concede (fail-closed). Serve a chi legge una copia locale o una
     * risposta di rete, dove il valore arriva come stringa.
     */
    public static Optional<AppRole> parse(String value) {
        if (value == null || value.isBlank()) {
            return Optional.empty();
        }
        try {
            return Optional.of(AppRole.valueOf(value));
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }
    }
}

File: frontend/packages/api-client/src/app-role.ts
// Ruolo della persona su una APPLICAZIONE (UC 0098/0099) e suo confronto, lato browser (UC 0101).
//
// Vive qui e non nel design system perché è parte del CONTRATTO di piattaforma, non della
// presentazione: il design system deve poter restare ignaro di che cosa sia un ruolo. Il gemello lato
// servizio è `app.appgrove.commons.access.AppRole` (Java), dove l'ordinamento è dichiarato una volta e
// nessuna applicazione lo riscrive. Qui l'ordinamento è dichiarato una seconda volta perché il browser
// non può leggere l'enumerazione Java — e una sola volta per lato, mai una terza.

/** I tre ruoli di applicazione, ordinati dal meno al più capace. L'ordine è la dichiarazione. */
export const APP_ROLES = ['viewer', 'editor', 'admin'] as const

export type AppRole = (typeof APP_ROLES)[number]

/**
 * Il ruolo posseduto è almeno tanto capace di quello richiesto?
 *
 * `role` accetta `null`/`undefined` per il caso «non lo sappiamo ancora» (il ruolo si legge dalla rete)
 * e per «nessun accesso all'applicazione»: in entrambi i casi la risposta è **no**. È deliberato e non
 * difensivo — in assenza di informazione non si concede, e un comando abilitato «in attesa» sarebbe una
 * promessa che il servizio poi smentisce con un rifiuto (UC 0101 §6).
 *
 * **L'owner dell'account non è un quarto valore**: la fonte di verità gli attribuisce `admin` su tutte le
 * applicazioni dell'account (UC 0098 §5), quindi passa ogni confronto senza che questa funzione sappia
 * che esiste. Un valore non riconosciuto non è un permesso.
 */
export function appRoleAtLeast(role: string | null | undefined, required: AppRole): boolean {
  if (!role) return false
  const held = APP_ROLES.indexOf(role as AppRole)
  if (held < 0) return false
  return held >= APP_ROLES.indexOf(required)
}
